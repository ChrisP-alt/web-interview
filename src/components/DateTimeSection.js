import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import Section from './Section'
import { ReactComponent as Icon } from '../icons/clock-regular.svg'
import styles from '../styles/components/dateTimeSection.module.scss'

const DateTimeSection = ({
  appointments,
  selectAppointment,
  selectedAppointment,
}) => {
  const buttons = appointments.map(appointment => {
    const isSelected =
      selectedAppointment && selectedAppointment.id === appointment.id
    const buttonClass = isSelected ? styles.buttonSelected : styles.button
    return (
      <button
        className={buttonClass}
        key={appointment.id}
        onClick={() => selectAppointment(appointment.id)}
        type="button"
      >
        {moment(appointment.time).format('DD/MM HH:mm')}
      </button>
    )
  })
  return (
    <Section headerText="Date & Time" Icon={Icon}>
      <div className={styles.wrapper}>{buttons}</div>
    </Section>
  )
}

// TODO: split into seperate custom PropType file
const appointmentPropType = PropTypes.shape({
  appointmentType: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  consultantType: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  id: PropTypes.number.isRequired,
  time: PropTypes.string.isRequired,
})

DateTimeSection.propTypes = {
  appointments: PropTypes.arrayOf(appointmentPropType),
  selectAppointment: PropTypes.func.isRequired,
  selectedAppointment: appointmentPropType,
}

export default DateTimeSection
