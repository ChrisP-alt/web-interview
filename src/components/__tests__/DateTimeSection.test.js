import React from 'react'
import { shallow } from 'enzyme'

import DateTimeSection from '../DateTimeSection'
import mockAvailableSlots from '../../__fixtures__/availableSlots'

describe('DateTimeSection component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <DateTimeSection
        appointments={mockAvailableSlots}
        selectAppointment={jest.fn()}
        selectedAppointment={mockAvailableSlots[0]}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders no buttonSelected when selectedAppointment is null', () => {
    const wrapper = shallow(
      <DateTimeSection
        appointments={mockAvailableSlots}
        selectAppointment={jest.fn()}
        selectedAppointment={null}
      />
    )
    expect(wrapper.find('.buttonSelected').length).toBe(0)
  })

  it('calls selectAppointment when a button is clicked', () => {
    const selectAppointment = jest.fn()
    const wrapper = shallow(
      <DateTimeSection
        appointments={mockAvailableSlots}
        selectAppointment={selectAppointment}
        selectedAppointment={null}
      />
    )
    const appointmentId = mockAvailableSlots[0].id
    const button = wrapper.findWhere(
      node => node.key() === appointmentId.toString()
    )
    button.simulate('click')
    expect(selectAppointment).toHaveBeenCalledWith(appointmentId)
  })
})
