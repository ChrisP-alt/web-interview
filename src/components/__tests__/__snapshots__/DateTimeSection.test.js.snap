// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeSection component renders correctly 1`] = `
<Section
  Icon={[Function]}
  headerText="Date & Time"
>
  <div
    className="wrapper"
  >
    <button
      className="buttonSelected"
      key="1"
      onClick={[Function]}
      type="button"
    >
      27/11 10:11
    </button>
    <button
      className="button"
      key="2"
      onClick={[Function]}
      type="button"
    >
      01/12 14:16
    </button>
    <button
      className="button"
      key="3"
      onClick={[Function]}
      type="button"
    >
      08/10 17:17
    </button>
  </div>
</Section>
`;
